### PR TITLE
Fix some debug glitches

### PR DESCRIPTION
**Pull request by dsyme**
I've been hunting for debug glitches and fixed a few:

1. https://github.com/dotnet/fsharp/issues/1390
2. https://github.com/dotnet/fsharp/issues/11977
3. https://github.com/dotnet/fsharp/issues/11978 
4. https://github.com/dotnet/fsharp/issues/11979

This builds on https://github.com/dotnet/fsharp/pull/11957 so compare here https://github.com/dsyme/fsharp/compare/bp5...dsyme:bp6

Baselines will need to be updated and some new debug testing added

### CODE REVIEW

**Comment by TIHan** [comment]
I think this PR includes https://github.com/dotnet/fsharp/pull/11957 - so that one should should be merged first before this one.

**Comment by dsyme** [comment]
This is now ready.  There's a separate issue for a case where the stepping on "when" patterns doesn't quite work with union patterns, see #11979, I'm going to look at that now but what we have in this PR is good to get in the bank.

In addition to the adjusted baselines and new tests, I've been through all of `tests\walkthroughs\DebugStepping\TheBigFileOfDebugStepping.fsx` manually and apart from the one glitch above can't see any issues.  


```
C:\GitHub\dsyme\fsharp>artifacts\bin\fsc\Debug\net472\fsc.exe -g --optimize- --tailcalls- --define:DEBUG tests\walkthroughs\DebugStepping\TheBigFileOfDebugStepping.fsx --langversion:preview
devenv /debugexe TheBigFileOfDebugStepping.exe
```

File: src/fsharp/PatternMatchCompilation.fs [hunk 1]
```diff
@@ -1434,3 +1422,5 @@ let rec CompilePattern  g denv amap tcVal infoReader exprm matchm warnOnUnused a
 
     | _ ->
         CompilePatternBasic g denv amap tcVal infoReader exprm matchm warnOnUnused true actionOnFailure (origInputVal, origInputValTypars, origInputExprOpt) clausesL inputTy resultTy
+type IA =
```
Comment (TIHan) [inline src/fsharp/PatternMatchCompilation.fs:1425]: I think this was an accident and should be removed.

**Review by TIHan — APPROVED** [review]
Change looks straight-forward and good to me.

Just a minor comment on some code that I think was accidental.

File: src/fsharp/PatternMatchCompilation.fs (hunk 1 above)
Comment (dsyme) [reply src/fsharp/PatternMatchCompilation.fs:1425]: Well spotted!

**Comment by dsyme** [comment]
> Just a minor comment on some code that I think was accidental.

I'll remove it in a separate PR

**Pull request merged**